[WIP] Partial fixes for cloud integration tests

michael-dev2rights opened the pull request:
##### SUMMARY

This is a set of fixes for the integration tests including minor improvements to error handling in ec2_key.

This is currently work in progress with more fixes to come.  

Partial fix for #22498

##### ISSUE TYPE

 - Bugfix Pull Request


##### COMPONENT NAME

cloud integration tests

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (mdd-integration-test-fix 8211ffb37f) last updated 2017/03/10 15:55:54 (GMT +100)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/mikedd/dev/ansible/ansible-add-modules/library']
  python version = 2.7.12 (default, Jul  1 2016, 15:12:24) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION



<!--- Paste verbatim command output below, e.g. before and after your change -->
```
verbatim output too big
```

mikedlr committed:
update expected message for unknown region to new one suggesting boto update

nerzhul commented on lib/ansible/modules/network/nsupdate.py:
```diff
@@ -40,7 +40,7 @@
 author: "Loic Blot (@nerzhul)"
 options:
-    state:
+    state:e
```
don't change my module please

michael-dev2rights replied:
@nerzhul sorry about that ; I somehow missed it checking the commit.  I have rebased off the most recent devel and at the same time dropped this from the changes.

mikedlr committed:
fixes to make test_ec2_eip and test_ec2_elb_lb work with current module interfaces

mikedlr committed:
longer timeouts make test_ec2_asg go further

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -370,7 +370,9 @@ def elb_healthy(asg_connection, elb_connection, module, group_name):
 
-            module.fail_json(msg=str(e))
+            module.fail_json(msg=str(e), exception=traceback.format_exc() )
+        except socket.gaierror as e:
```
I'm not sure about this. 
1) has socket been imported somewhere? 
2) It seems like a random place to include this error. What is the reason for it?

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -920,7 +922,11 @@ def main():
+            changed = delete_autoscaling_group(connection, module)
+        except boto.exception.BotoServerError as err:
+            module.fail_json(msg="AWS failed deleting autoscaling group - " + str(err),
+                             exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_elb_lb.py:
```diff
@@ -703,12 +708,17 @@ def _delete_elb(self):
+                                                          scheme=self.scheme)
+        except boto.exception.BotoServerError as err:
+            self.module.fail_json(msg="AWS rejected load balancer create operation - " + str(err),
+                                  exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_elb_lb.py:
```diff
@@ -963,7 +980,11 @@ def _set_access_log(self):
+        try:
+            attributes = self.elb.get_attributes()
+        except boto.exception.BotoServerError as err:
+            self.module.fail_json(msg="AWS rejected listing load balancer attributes - " + STR996(err),
```
STR996 intended to be str?

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_elb_lb.py:
```diff
@@ -963,7 +980,11 @@ def _set_access_log(self):
+            attributes = self.elb.get_attributes()
+        except boto.exception.BotoServerError as err:
+            self.module.fail_json(msg="AWS rejected listing load balancer attributes - " + STR996(err),
+                                  exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_elb_lb.py:
```diff
@@ -975,7 +996,11 @@ def _set_connection_draining_timeout(self):
+                self.elb_conn.modify_lb_attribute(self.name, 'ConnectionDraining', attributes.connection_draining)
+            except boto.exception.BotoServerError as err:
+                self.module.fail_json(msg="AWS rejected modifying attributes on LB - " + str(err),
+                                      exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -279,7 +279,12 @@ def main():
+        allgroups = ec2.get_all_security_groups()
+    except boto.exception.EC2ResponseError as err:
+        module.fail_json(msg="AWS rejected listing security groups - " + str(err),
+                         exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -325,7 +330,11 @@ def main():
+                    group = ec2.create_security_group(name, description, vpc_id=vpc_id)
+                except boto.exception.EC2ResponseError as err:
+                    module.fail_json(msg="AWS rejected security group creation - " + str(err),
+                                     exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -464,13 +473,17 @@ def main():
+                            cidr_ip=grant.cidr_ip)
+                    except boto.exception.EC2ResponseError as err:
+                        module.fail_json(msg="AWS rejected security group egress revocation - " + str(err),
+                         exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_key.py:
```diff
@@ -151,7 +151,11 @@ def main():
+        key = ec2.get_key_pair(name)
+    except boto.exception.EC2ResponseError as err:
+        module.fail_json(msg="AWS rejected key operation - " + str(err),
+                         exception=traceback.format_exc() )
```
There is a space before the closing parenthesis.

s-hertel commented on test/integration/roles/test_ec2_key/tasks/main.yml:
```diff
@@ -249,7 +249,7 @@
     state=present
     key_material='{{ item }}'
-  with_file: sshkey ~ '.pub'
+  with_file: '{{sshkey}}.pub'
```
Not sure I understand this.

michael-dev2rights replied:
sshkey is the location of the private key file which is created in the setup role.  {{sshkey}}.pub then becomes the public key which should be uploaded to AWS.

I have no idea if the original statement ever worked in some older version of Ansible.  I doubt it.

s-hertel commented:
The ec2_key tests including and after `TASK [test_ec2_key : test valid region parameter]` are still failing for me. This is really helpful progress though. Thanks for the time you've put into working on it.

s-hertel commented on test/integration/roles/test_ec2_key/tasks/main.yml:
```diff
@@ -50,7 +50,7 @@
+       - '"If the region definitely exists" in result.msg'
 
 # ============================================================
 - name: test valid region parameter
```
This tests succeeds for me so the following assert fails. This is the case for the next test `test environment variable EC2_REGION` as well. Are these tests failing appropriately here for you?

s-hertel commented on test/integration/roles/test_ecs_ecr/tasks/main.yml:
```diff
@@ -5,6 +5,10 @@
 - block:
     - name: When creating with check mode
       ecs_ecr: name='{{ ecr_name }}' region='{{ ec2_region }}'
+      environment:
```
For tests that don't specifically need environment variables to make sure they're working correctly can we make these global to eliminate the repetition?

michael-dev2rights replied:
my only fear here would be if the variables got accidentally inherited from the environment and then used someone's production account details.

willthames replied:
Use `profile: {{ profile|default(omit) }}`. 

I think people have to realise if they're running AWS tests that they're responsible for choosing which account it gets run against, but we need to provide a sensible mechanism for that, and profile is probably the best mechanism.

s-hertel replied:
@willthames Thanks for looking at this. Since we're repeating the credentials for every single task, I wish there was an easy way to extrapolate them. I don't want to set environment variables if we aren't specifically testing them (though changing every set of environment vars to a set of normal region, ec2_access_key, and ec2_secret_key variables feels kind of silly). I'm unsure how to approach this problem or how I feel about using profile. Our integration test [credentials.yml](https://github.com/ansible/ansible/blob/devel/test/integration/credentials.template) doesn't currently have an ec2_profile var so we would need to add that, update [documentation](https://github.com/ansible/ansible/blob/devel/test/integration/README.md#cloud-tests) and update all the tests. I guess we could also use the shell module to create a testing profile with the credentials.yml file and remove it afterward?

s-hertel replied:
@michael-dev2rights @willthames Retracting my thoughts on this. Until we've documented the process for switching to using profile we don't really have another choice. But we shouldn't use environment variables unless we're testing them specifically. Passing the credentials to the module is fine. I'll work on fixing those.

michael-dev2rights commented:
@s-hertel for the next few days I don't plan to do much here directly since I now have the integration tests working.  This would be a good opportunity for you to edit away in peace.  

As far as the environment goes, I think the crucial thing is to make sure that the credentials from the credentials.yml file are always used even if some other method, such as the configuration from the .aws directory has set up a different set of credentials.  The worst thing would be to have the integration tests disturb someone's production account.  

I think this means we should change to use either the *AWS* environment variables or direct module parameters in every case.  I haven't done this yet in this PR because I just wanted to do the minimum things to get the tests running.  

The commits are currently quite separated by theme.  It should be easy to cherry pick the fixes you think are easiest then we can make a separate PR later if needed.  If you need any help please just ask.  Best in IRC.

s-hertel committed:
fix ec2_asg integration tests with load balancer

add routing, create a security group, ensure to clean up launch configs afterward

fix indentation

mikedlr committed:
AWS integration tests - Create VPC, IGW, subnet and route table for use in other tests

mikedlr committed:
AWS integration tests / natgw - explicit region and other misc fixes

mikedlr committed:
add meta so that test_ec2_asg and test_ec2_vpc_nat_gateway can be run independently

mikedlr committed:
update ec2_eip, ec2_group, ec2_key and ec2_elb_lb integration tests to work with driving from profile

mikedlr committed:
aws integration tests - fixes to allow ec2 tests to run + move security group creation to setup

michael-dev2rights commented:
Currently the only failure that I am getting through running the complete set of AWS integration tests is in the ELB part of the test_ec2_asg tests.  

In order to run the tests I currently have to do some static security group setup which is documented in the README file (see commit ab1fda3f333d2c63bc78ecf391ee5125107e76c6).

s-hertel commented:
@michael-dev2rights What are the names of the ec2_asg tests that fail for you? Are you setting up the security group (which you've removed from test_ec2_asg for security reasons I'm sure) for the elbs with the vpc created in setup_ec2?
`test_ec2_asg : add 2 more instances wait for instances to be deemed healthy (ELB)` is failing for me, but I think the one before it fails too when your security group is set up incorrectly.

mikedlr committed:
automatic creation of security group rules to support ASG/LB tests

mikedlr committed:
aws integration tests - force public IP address assignment on instances for use in lb tests

michael-dev2rights commented:
@s-hertel with the logic I included in the commit just put up then I wasn't having any failures that I remember.

michael-dev2rights commented:
This is now obsolete because of the new way of doing integration tests, however there is still some work that needs to be cherry picked out.  I will delete once it's clear.

michael-dev2rights commented:
I have eliminated the stuff from this PR which was clearly obsolete.  I have also separated out fixes to modules.  This probably now contains useful work towards making these tests work, however I haven't tested it.

michael-dev2rights commented:
This pull request still has more working tests related to EC2 groups / load balancing groups / or similar which are working than the main test set.  It would be worth forward porting that, however it's not always obvious how to do that.  I've taken some of the valuable stuff and I am closing the PR for now to clean up my backlog and with the hope, but not likelyhood that someone will come back later and get some value from it.
